feat(android): percentage support in translate

intergalacticspacehighway opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory. -->

## Summary:
This PR adds percentage support in translate properties for android. Isolating this PR for easier reviews.

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

## Changelog:
[Android] [ADDED] - Percentage support in translate
<!-- Help reviewers and the release process by writing your own changelog entry.

Pick one each for the category and type tags:

[ANDROID|GENERAL|IOS|INTERNAL] [BREAKING|ADDED|CHANGED|DEPRECATED|REMOVED|FIXED|SECURITY] - Message

For more details, see:
https://reactnative.dev/contributing/changelogs-in-pull-requests
-->

## Test Plan:
- Checkout TransformExample.js -> Translate percentage example. 
- Added a simple test in `processTransform-test.js`. The regex is not perfect (values like 20px%, 20%px will pass, can be improved, let me know!)

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface. -->

intergalacticspacehighway commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/BaseViewManager.java:
```diff
@@ -177,18 +177,15 @@ public void setBackgroundColor(@NonNull T view, int backgroundColor) {
   public void setTransform(@NonNull T view, @Nullable ReadableArray matrix) {
     view.setTag(R.id.transform, matrix);
     view.setTag(R.id.invalidate_transform, true);
+    view.addOnLayoutChangeListener(this);
```
Yeah i remember discussing it here - https://github.com/facebook/react-native/pull/38558#discussion_r1271417406. To override `layout` i think we'll have to move transform logic to `ReactViewGroup`. 

`onLayout` runs during layout pass and can be used to update child view positions, in our case we update transform of the view. Not very ideal, i guess in the long run moving it to `RootViewGroup` might be better.

NickGerleman replied:
I checked code, and Android will no-op if adding existing listener, but it's a little bit smelly that we can add ourselves as a listener, after we have already added ourselves as listener. Not sure a better way though (probably don't want to waste cycles to always attach).

NickGerleman commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/BaseViewManager.java:
```diff
@@ -171,7 +171,7 @@ public void onLayoutChange(
       ReadableArray transformOrigin = (ReadableArray) v.getTag(R.id.transform_origin);
       ReadableArray transformMatrix = (ReadableArray) v.getTag(R.id.transform);
-      if (transformMatrix != null && transformOrigin != null) {
+      if (transformMatrix != null || transformOrigin != null) {
```
This isn't a matrix most of the time, right? Can we rename?

intergalacticspacehighway replied:
Agree, renamed to `transforms` since it's an array.

NickGerleman commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/uimanager/TransformHelper.java:
```diff
@@ -130,6 +152,14 @@ public static void processTransform(
   }
 
+  private static double parsePercentageToValue(String stringValue, double dimension) {
+    if (stringValue.endsWith("%")) {
```
I'm not sure the old code was reliable here, but we usually want to avoid native crashes on invalid style values. So we probably want to be handling `NumberFormatException` somewhere.

intergalacticspacehighway replied:
makes sense, updated!

NickGerleman reviewed (commented):
Left a couple of minor comments, but this seems pretty nicely scoped 👍

intergalacticspacehighway commented:
> I checked code, and Android will no-op if adding existing listener, but it's a little bit smelly that we can add ourselves as a listener, after we have already added ourselves as listener. Not sure a better way though (probably don't want to waste cycles to always attach).

i moved it in [onAfterUpdateTransaction](https://github.com/facebook/react-native/pull/43193/commits/5e705110ba1b26b76b9ca33f1d67b81c9725faba) so it gets called a bit less and yes duplicate calls for same listener will be ignored. But i agree this is not very optimal. If all views are getting extended from `ReactViewGroup` then we can move the entire transform logic there as it already has a `onSizeCallback` as mentioned [here](https://github.com/facebook/react-native/pull/38558#discussion_r1271417406).
